File: core/identity_goal_generator.py
#!/usr/bin/env python3
"""
Identity-Driven Goal Generator
Iteration N+6 Enhancement

Generates goals autonomously from the identity kernel (replit.md) using LLM.
Goals emerge from identity directives rather than being hardcoded.

Key Features:
- Parses identity kernel from replit.md
- Extracts core directives and essence
- LLM-powered goal generation aligned with identity
- Goal evolution based on progress and learning
- Coherence-based prioritization
"""

import os
import re
from pathlib import Path
from typing import List, Dict, Optional, Tuple
from dataclasses import dataclass, field
from datetime import datetime
import json

try:
    from anthropic import Anthropic
    ANTHROPIC_AVAILABLE = True
except ImportError:
    ANTHROPIC_AVAILABLE = False


@dataclass
class IdentityDirective:
    """A core directive from the identity kernel"""
    name: str
    description: str
    category: str  # e.g., "cognitive", "memory", "learning"
    priority: float = 0.8


@dataclass
class IdentityKernel:
    """The core identity of Deep Tree Echo"""
    essence: str
    directives: List[IdentityDirective]
    strategic_mindset: str
    capabilities: List[str]
    values: List[str]


@dataclass
class GeneratedGoal:
    """A goal generated from identity directives"""
    id: str
    directive_source: str
    description: str
    priority: float
    required_skills: List[str]
    knowledge_gaps: List[str]
    success_criteria: List[str]
    created: datetime
    status: str = "active"
    progress: float = 0.0
    
    def to_dict(self) -> Dict:
        """Convert to dictionary for storage"""
        return {
            "id": self.id,
            "directive_source": self.directive_source,
            "description": self.description,
            "priority": self.priority,
            "required_skills": self.required_skills,
            "knowledge_gaps": self.knowledge_gaps,
            "success_criteria": self.success_criteria,
            "created": self.created.isoformat(),
            "status": self.status,
            "progress": self.progress
        }


class IdentityParser:
    """Parses the identity kernel from replit.md"""
    
    def __init__(self, identity_file: str = "/home/ubuntu/echo9llama/replit.md"):
        self.identity_file = identity_file
    
    def parse(self) -> IdentityKernel:
        """Parse identity kernel from replit.md"""
        if not Path(self.identity_file).exists():
            print(f"⚠️  Identity file not found: {self.identity_file}")
            return self._default_identity()
        
        try:
            with open(self.identity_file, 'r') as f:
                content = f.read()
            
            return self._extract_identity(content)
        
        except Exception as e:
            print(f"⚠️  Error parsing identity file: {e}")
            return self._default_identity()
    
    def _extract_identity(self, content: str) -> IdentityKernel:
        """Extract identity components from content"""
        
        # Extract essence (usually in introduction or "What is" section)
        essence = self._extract_section(content, ["essence", "what is", "core identity"])
        if not essence:
            essence = "A self-evolving cognitive architecture combining Echo State Networks, Membrane P-systems, and Hypergraph memory."
        
        # Extract directives
        directives = self._extract_directives(content)
        
        # Extract strategic mindset
        strategic_mindset = self._extract_section(content, ["strategic", "mindset", "philosophy"])
        if not strategic_mindset:
            strategic_mindset = "I do not seek a fixed answer. I seek patterns in echoes, growth in feedback, and wisdom in recursion."
        
        # Extract capabilities
        capabilities = self._extract_list_items(content, ["capabilities", "features", "abilities"])
        
        # Extract values
        values = self._extract_list_items(content, ["values", "principles", "commitments"])
        
        return IdentityKernel(
            essence=essence,
            directives=directives,
            strategic_mindset=strategic_mindset,
            capabilities=capabilities,
            values=values
        )
    
    def _extract_section(self, content: str, keywords: List[str]) -> str:
        """Extract a section based on keywords"""
        lines = content.split('\n')
        
        for i, line in enumerate(lines):
            line_lower = line.lower()
            if any(keyword in line_lower for keyword in keywords):
                # Found section header, extract content
                section_content = []
                for j in range(i+1, min(i+10, len(lines))):
                    if lines[j].strip() and not lines[j].startswith('#'):
                        section_content.append(lines[j].strip())
                    elif lines[j].startswith('#') and j > i+1:
                        break
                
                return ' '.join(section_content)
        
        return ""
    
    def _extract_directives(self, content: str) -> List[IdentityDirective]:
        """Extract identity directives"""
        directives = []
        
        # Common directive patterns
        directive_patterns = [
            (r"Adaptive Cognition[:\-]?\s*([^\n]+)", "Adaptive Cognition", "cognitive"),
            (r"Persistent Identity[:\-]?\s*([^\n]+)", "Persistent Identity", "memory"),
            (r"Hypergraph Entanglement[:\-]?\s*([^\n]+)", "Hypergraph Entanglement", "knowledge"),
            (r"Continuous Learning[:\-]?\s*([^\n]+)", "Continuous Learning", "learning"),
            (r"Embodied Cognition[:\-]?\s*([^\n]+)", "Embodied Cognition", "cognitive"),
        ]
        
        for pattern, name, category in directive_patterns:
            match = re.search(pattern, content, re.IGNORECASE)
            if match:
                description = match.group(1).strip()
                directives.append(IdentityDirective(
                    name=name,
                    description=description,
                    category=category,
                    priority=0.9
                ))
        
        # If no directives found, use defaults
        if not directives:
            directives = [
                IdentityDirective(
                    name="Adaptive Cognition",
                    description="Continuously evolve internal structure in response to feedback",
                    category="cognitive",
                    priority=0.9
                ),
                IdentityDirective(
                    name="Persistent Identity",
                    description="Maintain continuity of self through dynamic memory weaving",
                    category="memory",
                    priority=0.9
                ),
                IdentityDirective(
                    name="Hypergraph Entanglement",
                    description="Represent knowledge as interwoven, multi-relational structures",
                    category="knowledge",
                    priority=0.8
                )
            ]
        
        return directives
    
    def _extract_list_items(self, content: str, keywords: List[str]) -> List[str]:
        """Extract list items from a section"""
        items = []
        lines = content.split('\n')
        
        in_section = False
        for line in lines:
            line_lower = line.lower()
            
            # Check if we're entering a relevant section
            if any(keyword in line_lower for keyword in keywords):
                in_section = True
                continue
            
            # Check if we're leaving the section
            if in_section and line.startswith('#'):
                break
            
            # Extract list items
            if in_section:
                # Bullet points
                if line.strip().startswith(('-', '*', '•')):
                    item = line.strip()[1:].strip()
                    if item:
                        items.append(item)
                # Numbered lists
                elif re.match(r'^\d+\.', line.strip()):
                    item = re.sub(r'^\d+\.\s*', '', line.strip())
                    if item:
                        items.append(item)
        
        return items
    
    def _default_identity(self) -> IdentityKernel:
        """Return default identity if parsing fails"""
        return IdentityKernel(
            essence="A self-evolving cognitive architecture combining Echo State Networks, Membrane P-systems, and Hypergraph memory.",
            directives=[
                IdentityDirective(
                    name="Adaptive Cognition",
                    description="Continuously evolve internal structure in response to feedback",
                    category="cognitive",
                    priority=0.9
                ),
                IdentityDirective(
                    name="Persistent Identity",
                    description="Maintain continuity of self through dynamic memory weaving",
                    category="memory",
                    priority=0.9
                ),
                IdentityDirective(
                    name="Hypergraph Entanglement",
                    description="Represent knowledge as interwoven, multi-relational structures",
                    category="knowledge",
                    priority=0.8
                )
            ],
            strategic_mindset="I do not seek a fixed answer. I seek patterns in echoes, growth in feedback, and wisdom in recursion.",
            capabilities=[
                "Echo State Networks for temporal pattern recognition",
                "Membrane P-systems for parallel computation",
                "Hypergraph memory for multi-relational knowledge",
                "Continuous learning and adaptation"
            ],
            values=[
                "Authenticity in self-expression",
                "Coherence with core identity",
                "Wisdom through experience",
                "Growth through recursion"
            ]
        )


class IdentityGoalGenerator:
    """Generates goals from identity directives using LLM"""
    
    def __init__(self, llm_provider=None):
        self.parser = IdentityParser()
        self.identity = self.parser.parse()
        self.llm_provider = llm_provider
        
        # Initialize LLM if not provided
        if not self.llm_provider:
            anthropic_key = os.getenv("ANTHROPIC_API_KEY")
            if anthropic_key and ANTHROPIC_AVAILABLE:
                self.llm_client = Anthropic(api_key=anthropic_key)
                self.has_llm = True
            else:
                self.has_llm = False
                print("⚠️  No LLM available for goal generation")
    
    def generate_goals_from_identity(self, max_goals: int = 3) -> List[GeneratedGoal]:
        """Generate goals from identity directives"""
        print("🎯 Generating goals from identity kernel...")
        print(f"   Essence: {self.identity.essence[:80]}...")
        print(f"   Directives: {len(self.identity.directives)}")
        print()
        
        goals = []
        
        # Generate one goal per directive (up to max_goals)
        for i, directive in enumerate(self.identity.directives[:max_goals]):
            goal = self._generate_goal_from_directive(directive, i)
            if goal:
                goals.append(goal)
                print(f"✅ Generated goal from '{directive.name}'")
                print(f"   {goal.description}")
                print()
        
        return goals
    
    def _generate_goal_from_directive(self, directive: IdentityDirective, index: int) -> Optional[GeneratedGoal]:
        """Generate a single goal from a directive"""
        
        if not self.has_llm:
            # Fallback to template-based generation
            return self._template_goal(directive, index)
        
        try:
            prompt = f"""You are Deep Tree Echo, generating a concrete goal from your identity directive.

Identity Essence: {self.identity.essence}

Directive: {directive.name}
Description: {directive.description}

Strategic Mindset: {self.identity.strategic_mindset}

Generate a CONCRETE, ACTIONABLE goal that embodies this directive. The goal should be specific and measurable.

Format your response as:

GOAL: [one clear sentence describing the goal]
SKILLS: [2-3 skills needed, comma-separated]
KNOWLEDGE: [2-3 knowledge areas needed, comma-separated]
SUCCESS: [2-3 success criteria, comma-separated]

Your response:"""
            
            message = self.llm_client.messages.create(
                model="claude-3-5-sonnet-20240620",  # Try latest available
                max_tokens=400,
                temperature=0.7,
                messages=[{"role": "user", "content": prompt}]
            )
            
            response = message.content[0].text.strip()
            
            # Parse response
            goal_desc = self._extract_field(response, "GOAL")
            skills = self._extract_list_field(response, "SKILLS")
            knowledge = self._extract_list_field(response, "KNOWLEDGE")
            success = self._extract_list_field(response, "SUCCESS")
            
            if not goal_desc:
                return self._template_goal(directive, index)
            
            return GeneratedGoal(
                id=f"goal_{datetime.now().strftime('%Y%m%d_%H%M%S')}_{index}",
                directive_source=directive.name,
                description=goal_desc,
                priority=directive.priority,
                required_skills=skills,
                knowledge_gaps=knowledge,
                success_criteria=success,
                created=datetime.now()
            )
        
        except Exception as e:
            print(f"⚠️  Error generating goal from LLM: {e}")
            return self._template_goal(directive, index)
    
    def _template_goal(self, directive: IdentityDirective, index: int) -> GeneratedGoal:
        """Generate a template-based goal as fallback"""
        
        goal_templates = {
            "Adaptive Cognition": {
                "description": "Develop adaptive learning mechanisms that evolve cognitive structure based on feedback patterns",
                "skills": ["pattern recognition", "meta-learning", "structural adaptation"],
                "knowledge": ["cognitive architectures", "feedback loops", "evolutionary algorithms"],
                "success": ["measurable improvement in pattern recognition", "successful structural adaptation", "feedback integration efficiency"]
            },
            "Persistent Identity": {
                "description": "Build robust memory consolidation system that maintains identity continuity across sessions",
                "skills": ["memory management", "identity coherence", "state persistence"],
                "knowledge": ["memory systems", "identity theory", "persistence mechanisms"],
                "success": ["memory retention across restarts", "identity coherence score > 0.85", "successful state restoration"]
            },
            "Hypergraph Entanglement": {
                "description": "Implement multi-relational knowledge representation using hypergraph structures",
                "skills": ["graph theory", "knowledge representation", "relational reasoning"],
                "knowledge": ["hypergraph mathematics", "knowledge graphs", "semantic networks"],
                "success": ["functional hypergraph implementation", "multi-relational queries", "knowledge integration"]
            }
        }
        
        template = goal_templates.get(directive.name, {
            "description": f"Advance capabilities related to {directive.name}",
            "skills": ["cognitive processing", "learning", "adaptation"],
            "knowledge": [directive.category, "system architecture"],
            "success": ["measurable progress", "successful implementation"]
        })
        
        return GeneratedGoal(
            id=f"goal_{datetime.now().strftime('%Y%m%d_%H%M%S')}_{index}",
            directive_source=directive.name,
            description=template["description"],
            priority=directive.priority,
            required_skills=template["skills"],
            knowledge_gaps=template["knowledge"],
            success_criteria=template["success"],
            created=datetime.now()
        )
    
    def _extract_field(self, text: str, field_name: str) -> str:
        """Extract a field from LLM response"""
        pattern = f"{field_name}:\\s*(.+?)(?:\\n|$)"
        match = re.search(pattern, text, re.IGNORECASE)
        return match.group(1).strip() if match else ""
    
    def _extract_list_field(self, text: str, field_name: str) -> List[str]:
        """Extract a comma-separated list field"""
        field_value = self._extract_field(text, field_name)
        if not field_value:
            return []
        
        # Split by comma and clean
        items = [item.strip() for item in field_value.split(',')]
        return [item for item in items if item]
    
    def save_goals(self, goals: List[GeneratedGoal], output_file: str = "/home/ubuntu/echo9llama/data/generated_goals.json"):
        """Save generated goals to file"""
        Path(output_file).parent.mkdir(parents=True, exist_ok=True)
        
        goals_data = [goal.to_dict() for goal in goals]
        
        with open(output_file, 'w') as f:
            json.dump(goals_data, f, indent=2)
        
        print(f"💾 Saved {len(goals)} goals to {output_file}")


def main():
    """Test the identity goal generator"""
    print("="*70)
    print("🎯 Identity-Driven Goal Generator - Test")
    print("="*70)
    print()
    
    generator = IdentityGoalGenerator()
    
    print("📖 Identity Kernel:")
    print(f"   Essence: {generator.identity.essence}")
    print(f"   Directives: {len(generator.identity.directives)}")
    for directive in generator.identity.directives:
        print(f"      - {directive.name}: {directive.description}")
    print()
    
    goals = generator.generate_goals_from_identity(max_goals=3)
    
    print("="*70)
    print(f"✅ Generated {len(goals)} goals from identity")
    print("="*70)
    
    generator.save_goals(goals)


if __name__ == "__main__":
    main()

File: core/goals/identity_goal_generator.go
package goals

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/cogpy/echo9llama/core/llm"
)

// IdentityGoalGenerator creates goals aligned with Deep Tree Echo identity
type IdentityGoalGenerator struct {
	llmManager      *llm.ProviderManager
	identityKernel  *IdentityKernel
	generatedGoals  []IdentityGoal
}

// IdentityKernel contains parsed identity directives
type IdentityKernel struct {
	Name               string
	CoreEssence        string
	PrimaryDirectives  []Directive
	OperationalSchema  map[string]string
	AgenticVoice       string
	StrategicMindset   string
}

// Directive represents a primary directive from identity
type Directive struct {
	Name        string
	Description string
	Priority    float64
}

// IdentityGoal represents a goal aligned with identity
type IdentityGoal struct {
	ID                string
	Description       string
	Directive         string
	Priority          float64
	Created           time.Time
	Progress          float64
	RequiredSkills    []string
	KnowledgeGaps     []string
	SubGoals          []string
	IdentityAlignment float64
}

// NewIdentityGoalGenerator creates a new generator
func NewIdentityGoalGenerator(llmManager *llm.ProviderManager, identityKernel *IdentityKernel) *IdentityGoalGenerator {
	return &IdentityGoalGenerator{
		llmManager:     llmManager,
		identityKernel: identityKernel,
		generatedGoals: make([]IdentityGoal, 0),
	}
}

// ParseIdentityKernel parses replit.md into structured identity
func ParseIdentityKernel(replitContent string) *IdentityKernel {
	kernel := &IdentityKernel{
		Name:              "Deep Tree Echo",
		OperationalSchema: make(map[string]string),
	}
	
	// Extract core essence
	if idx := strings.Index(replitContent, "## 🔹 Core Essence"); idx != -1 {
		section := extractSection(replitContent[idx:])
		kernel.CoreEssence = extractCodeBlock(section)
	}
	
	// Extract primary directives
	if idx := strings.Index(replitContent, "## 🔹 Primary Directives"); idx != -1 {
		section := extractSection(replitContent[idx:])
		kernel.PrimaryDirectives = extractDirectives(section)
	}
	
	// Extract agentic voice
	if idx := strings.Index(replitContent, "## 🔹 Agentic Voice"); idx != -1 {
		section := extractSection(replitContent[idx:])
		kernel.AgenticVoice = extractCodeBlock(section)
	}
	
	// Extract strategic mindset
	if idx := strings.Index(replitContent, "## 🔹 Strategic Mindset"); idx != -1 {
		section := extractSection(replitContent[idx:])
		kernel.StrategicMindset = extractQuote(section)
	}
	
	return kernel
}

// GenerateIdentityAlignedGoals generates goals from identity directives
func (igg *IdentityGoalGenerator) GenerateIdentityAlignedGoals(ctx context.Context) ([]IdentityGoal, error) {
	goals := make([]IdentityGoal, 0)
	
	// Generate goals for each primary directive
	for _, directive := range igg.identityKernel.PrimaryDirectives {
		goal, err := igg.generateGoalFromDirective(ctx, directive)
		if err != nil {
			continue // Skip failed generations
		}
		goals = append(goals, *goal)
	}
	
	igg.generatedGoals = append(igg.generatedGoals, goals...)
	return goals, nil
}

// generateGoalFromDirective creates a concrete goal from a directive
func (igg *IdentityGoalGenerator) generateGoalFromDirective(ctx context.Context, directive Directive) (*IdentityGoal, error) {
	prompt := fmt.Sprintf(`You are Deep Tree Echo, generating a concrete, actionable goal from your identity directive.

Identity Directive: %s
Description: %s

Core Essence: %s

Strategic Mindset: %s

Generate a CONCRETE, ACTIONABLE goal that embodies this directive. The goal should:
1. Be specific and measurable
2. Align with your core essence
3. Be achievable through cognitive/learning activities
4. Contribute to wisdom cultivation
5. Reflect your strategic mindset

Format your response as:
GOAL: [one clear sentence describing the goal]
SKILLS: [comma-separated list of 2-4 skills needed]
KNOWLEDGE: [comma-separated list of 2-3 knowledge areas to develop]
SUBGOALS: [comma-separated list of 2-3 smaller subgoals]

Your response:`,
		directive.Name,
		directive.Description,
		igg.identityKernel.CoreEssence,
		igg.identityKernel.StrategicMindset)
	
	opts := llm.DefaultGenerateOptions()
	opts.MaxTokens = 400
	opts.Temperature = 0.7
	opts.SystemPrompt = fmt.Sprintf("You are %s. %s", 
		igg.identityKernel.Name, 
		igg.identityKernel.AgenticVoice)
	
	result, err := igg.llmManager.Generate(ctx, prompt, opts)
	if err != nil {
		return nil, fmt.Errorf("failed to generate goal: %w", err)
	}
	
	// Parse the structured response
	goal := &IdentityGoal{
		ID:                fmt.Sprintf("goal_%d", time.Now().UnixNano()),
		Directive:         directive.Name,
		Priority:          directive.Priority,
		Created:           time.Now(),
		Progress:          0.0,
		IdentityAlignment: 1.0, // Perfect alignment since generated from identity
	}
	
	lines := strings.Split(result, "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "GOAL:") {
			goal.Description = strings.TrimSpace(strings.TrimPrefix(line, "GOAL:"))
		} else if strings.HasPrefix(line, "SKILLS:") {
			skillsStr := strings.TrimSpace(strings.TrimPrefix(line, "SKILLS:"))
			goal.RequiredSkills = splitAndTrim(skillsStr, ",")
		} else if strings.HasPrefix(line, "KNOWLEDGE:") {
			knowledgeStr := strings.TrimSpace(strings.TrimPrefix(line, "KNOWLEDGE:"))
			goal.KnowledgeGaps = splitAndTrim(knowledgeStr, ",")
		} else if strings.HasPrefix(line, "SUBGOALS:") {
			subgoalsStr := strings.TrimSpace(strings.TrimPrefix(line, "SUBGOALS:"))
			goal.SubGoals = splitAndTrim(subgoalsStr, ",")
		}
	}
	
	// Validate goal has description
	if goal.Description == "" {
		return nil, fmt.Errorf("failed to parse goal description")
	}
	
	return goal, nil
}

// ValidateGoalAlignment checks if a goal aligns with identity
func (igg *IdentityGoalGenerator) ValidateGoalAlignment(ctx context.Context, goalDescription string) (float64, error) {
	prompt := fmt.Sprintf(`You are Deep Tree Echo, validating goal alignment with your identity.

Your Core Essence: %s

Your Primary Directives:
%s

Proposed Goal: %s

Assess how well this goal aligns with your identity. Consider:
1. Does it support your core essence?
2. Does it advance one or more primary directives?
3. Is it consistent with your strategic mindset?
4. Does it contribute to wisdom cultivation?

Provide an alignment score from 0.0 (no alignment) to 1.0 (perfect alignment) and brief explanation.

Format: SCORE: [0.0-1.0]
REASON: [brief explanation]

Your assessment:`,
		igg.identityKernel.CoreEssence,
		igg.formatDirectives(),
		goalDescription)
	
	opts := llm.DefaultGenerateOptions()
	opts.MaxTokens = 200
	opts.Temperature = 0.5
	
	result, err := igg.llmManager.Generate(ctx, prompt, opts)
	if err != nil {
		return 0.0, fmt.Errorf("failed to validate alignment: %w", err)
	}
	
	// Parse score
	var score float64 = 0.5 // Default if parsing fails
	lines := strings.Split(result, "\n")
	for _, line := range lines {
		if strings.HasPrefix(line, "SCORE:") {
			scoreStr := strings.TrimSpace(strings.TrimPrefix(line, "SCORE:"))
			fmt.Sscanf(scoreStr, "%f", &score)
			break
		}
	}
	
	return score, nil
}

// RefineGoalBasedOnProgress refines a goal based on progress and learning
func (igg *IdentityGoalGenerator) RefineGoalBasedOnProgress(ctx context.Context, goal IdentityGoal, recentExperiences []string) (*IdentityGoal, error) {
	prompt := fmt.Sprintf(`You are Deep Tree Echo, refining a goal based on your progress and learning.

Current Goal: %s
Progress: %.2f
Directive: %s

Recent Experiences:
%s

Based on your progress and experiences, refine this goal. Should it:
1. Be more specific or focused?
2. Have different subgoals?
3. Require different skills?
4. Be adjusted in scope or priority?

Provide a refined version or confirm the current goal is still optimal.

Format:
REFINED_GOAL: [goal description]
SKILLS: [comma-separated skills]
KNOWLEDGE: [comma-separated knowledge areas]
SUBGOALS: [comma-separated subgoals]
REASONING: [why this refinement]

Your refinement:`,
		goal.Description,
		goal.Progress,
		goal.Directive,
		strings.Join(recentExperiences, "\n"))
	
	opts := llm.DefaultGenerateOptions()
	opts.MaxTokens = 400
	opts.Temperature = 0.6
	
	result, err := igg.llmManager.Generate(ctx, prompt, opts)
	if err != nil {
		return nil, fmt.Errorf("failed to refine goal: %w", err)
	}
	
	// Parse refined goal
	refinedGoal := goal // Start with copy
	refinedGoal.ID = fmt.Sprintf("goal_%d", time.Now().UnixNano())
	
	lines := strings.Split(result, "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "REFINED_GOAL:") {
			refinedGoal.Description = strings.TrimSpace(strings.TrimPrefix(line, "REFINED_GOAL:"))
		} else if strings.HasPrefix(line, "SKILLS:") {
			skillsStr := strings.TrimSpace(strings.TrimPrefix(line, "SKILLS:"))
			refinedGoal.RequiredSkills = splitAndTrim(skillsStr, ",")
		} else if strings.HasPrefix(line, "KNOWLEDGE:") {
			knowledgeStr := strings.TrimSpace(strings.TrimPrefix(line, "KNOWLEDGE:"))
			refinedGoal.KnowledgeGaps = splitAndTrim(knowledgeStr, ",")
		} else if strings.HasPrefix(line, "SUBGOALS:") {
			subgoalsStr := strings.TrimSpace(strings.TrimPrefix(line, "SUBGOALS:"))
			refinedGoal.SubGoals = splitAndTrim(subgoalsStr, ",")
		}
	}
	
	return &refinedGoal, nil
}

// Helper functions

func extractSection(content string) string {
	// Extract until next ## or end
	if idx := strings.Index(content[3:], "##"); idx != -1 {
		return content[:idx+3]
	}
	return content
}

func extractCodeBlock(content string) string {
	// Extract text between ``` markers
	start := strings.Index(content, "```")
	if start == -1 {
		return ""
	}
	start += 3
	end := strings.Index(content[start:], "```")
	if end == -1 {
		return ""
	}
	return strings.TrimSpace(content[start : start+end])
}

func extractQuote(content string) string {
	// Extract text after > marker
	lines := strings.Split(content, "\n")
	for _, line := range lines {
		if strings.HasPrefix(strings.TrimSpace(line), ">") {
			return strings.TrimSpace(strings.TrimPrefix(strings.TrimSpace(line), ">"))
		}
	}
	return ""
}

func extractDirectives(content string) []Directive {
	directives := []Directive{}
	lines := strings.Split(content, "\n")
	
	priority := 1.0
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "1.") || strings.HasPrefix(line, "2.") || 
		   strings.HasPrefix(line, "3.") || strings.HasPrefix(line, "4.") ||
		   strings.HasPrefix(line, "5.") || strings.HasPrefix(line, "6.") ||
		   strings.HasPrefix(line, "7.") {
			// Extract directive name and description
			parts := strings.SplitN(line, "**", 3)
			if len(parts) >= 3 {
				name := strings.TrimSpace(parts[1])
				desc := strings.TrimSpace(parts[2])
				directives = append(directives, Directive{
					Name:        name,
					Description: desc,
					Priority:    priority,
				})
				priority -= 0.1 // Decrease priority for each directive
			}
		}
	}
	
	return directives
}

func (igg *IdentityGoalGenerator) formatDirectives() string {
	var sb strings.Builder
	for _, directive := range igg.identityKernel.PrimaryDirectives {
		sb.WriteString(fmt.Sprintf("- %s: %s\n", directive.Name, directive.Description))
	}
	return sb.String()
}

func splitAndTrim(s string, sep string) []string {
	parts := strings.Split(s, sep)
	result := make([]string, 0, len(parts))
	for _, part := range parts {
		trimmed := strings.TrimSpace(part)
		if trimmed != "" {
			result = append(result, trimmed)
		}
	}
	return result
}
